packages/squiggle-lang/src/public/SqProject/SqModule.ts

```diff
@@ -0,0 +1,180 @@
+import { parse } from "../../ast/parse.js";
+import { AST, LocationRange } from "../../ast/types.js";
+import { Env } from "../../dists/env.js";
+import { errMap, result } from "../../utility/result.js";
+import {
+  SqCompileError,
+  SqError,
+  SqImportError,
+  SqOtherError,
+} from "../SqError.js";
+import { SqLinker } from "../SqLinker.js";
+import { ModulePointer, ProjectState } from "./ProjectState.js";
+import { SqModuleOutput } from "./SqModuleOutput.js";
+import { getHash } from "./utils.js";
+
+export type Import = {
+  name: string;
+  hash: string | undefined;
+  variable: string;
+  location: LocationRange;
+};
+
+export function importToPointer(importBinding: Import): ModulePointer {
+  return {
+    name: importBinding.name,
+    hash: importBinding.hash,
+  };
+}
+
+type ImportModules =
+  | {
+      type: "loaded";
+      value: Record<string, SqModule>;
+    }
+  | {
+      type: "loading";
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+type ImportOutputs =
+  | {
+      type: "loaded";
+      value: Record<string, SqModuleOutput>;
+    }
+  | {
+      type: "loading"; // or running - i.e. ImportModules might be loaded, but outputs don't exist yet
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+export class SqModule {
+  name: string;
+  code: string;
+  // key is module name, value is hash
+  pins: Record<string, string>;
+
+  private _ast?: result<AST, SqError>;
+
+  constructor(params: {
+    name: string;
+    code: string;
+    pins?: Record<string, string>;
+  }) {
+    this.name = params.name;
+    this.code = params.code;
+    this.pins = params.pins ?? {};
+  }
+
+  // For now, parsing is done lazily but synchronously and on happens on the
+  // main thread. Parsing is usually fast enough and this makes the
+  // implementation simpler.
+  ast() {
+    if (!this._ast) {
+      this._ast = errMap(
+        parse(this.code, this.name),
+        (e) => new SqCompileError(e)
+      );
+    }
+    return this._ast;
+  }
+
+  getImports(linker: SqLinker): Import[] {
```

Comment:
Having to pass `linker` here is annoying.

But it's necessary because of `linker.resolve`, which is needed in CLI (relative -> absolute paths).